extensions-core/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerFactory.java

```diff
@@ -70,22 +72,44 @@ public KubernetesTaskRunnerFactory(
   @Override
   public KubernetesTaskRunner build()
   {
-    KubernetesPeonClient peonClient;
-    if (adapterTypeAllowingTasksInDifferentNamespaces.contains(taskAdapter.getAdapterType())) {
-      peonClient = new KubernetesPeonClient(
-          druidKubernetesClient,
-          kubernetesTaskRunnerConfig.getNamespace(),
-          kubernetesTaskRunnerConfig.getOverlordNamespace(),
-          kubernetesTaskRunnerConfig.isDebugJobs(),
-          emitter
-      );
+    AbstractKubernetesPeonClient peonClient;
+    boolean enableCache = kubernetesTaskRunnerConfig.isEnableKubernetesClientSharedInformers();
+    boolean useOverlordNamespace = adapterTypeAllowingTasksInDifferentNamespaces.contains(taskAdapter.getAdapterType());
+
+    if (enableCache) {
+      if (useOverlordNamespace) {
+        peonClient = new CachingKubernetesPeonClient(
+            druidKubernetesClient,
+            kubernetesTaskRunnerConfig.getNamespace(),
+            kubernetesTaskRunnerConfig.getOverlordNamespace(),
+            kubernetesTaskRunnerConfig.isDebugJobs(),
+            emitter
+        );
+      } else {
+        peonClient = new CachingKubernetesPeonClient(
```

Comment:
For brevity, let's just use a single constructor and pass the overlord namespace as null when `useOverlordNamespace` is `false`. The constructor can then default it to empty string or leave null, whichever is appropriate.